File: shared/units.ts
// Data-driven unit content: stats + presentation for every trainable unit. New
// unit types slot in here by adding one UNIT_DEFS entry — the generic combat,
// gather and movement systems already dispatch on the numeric kind, so no
// systems code changes when the roster grows.

import {
  UnitKind,
  BuildingKind,
  DamageType,
  ArmorClass,
} from './enums.ts';
import type { ResourceCost } from './economy.ts';

export interface UnitDef {
  label: string;
  icon: string; // emoji shown in the train/selection UI
  speed: number; // world units / second
  carry: number; // wood per gather trip (0 = non-gatherer)
  radius: number;
  height: number;
  maxHp: number;
  attack: number; // base damage per hit (0 = non-combatant)
  damageType: DamageType;
  armorClass: ArmorClass;
  bonusVsArmor?: Partial<Record<ArmorClass, number>>; // specialist counter
  range: number; // attack reach in world units
  attackRate: number; // seconds between hits
  aggroRange: number; // auto-acquire enemies within (0 = never auto-aggro)
  cost: ResourceCost; // resources to train
  tint?: number; // mesh tint override; otherwise owner color
  requires?: BuildingKind; // extra tech prereq beyond the training building
  prefersBuildings?: boolean; // siege: hunt structures over soft targets
}

export const UNIT_DEFS: Record<UnitKind, UnitDef> = {
  [UnitKind.Peasant]: {
    label: 'Peasant',
    icon: '🧑‍🌾',
    speed: 2.5,
    carry: 8,
    radius: 0.22,
    height: 0.7,
    maxHp: 30,
    attack: 0,
    damageType: DamageType.Blunt,
    armorClass: ArmorClass.Unarmored,
    range: 0.8,
    attackRate: 1.2,
    aggroRange: 0,
    cost: { wood: 20 },
  },
  [UnitKind.Spearman]: {
    label: 'Spearman',
    icon: '🛡️',
    speed: 2.2,
    carry: 0,
    radius: 0.26,
    height: 0.85,
    maxHp: 70,
    attack: 12,
    damageType: DamageType.Pierce,
    armorClass: ArmorClass.Leather,
    bonusVsArmor: { [ArmorClass.Mail]: 2.6 }, // braced against mailed cavalry
    range: 1.2, // long reach — outranges a knight's blade
    attackRate: 1.0,
    aggroRange: 6,
    cost: { wood: 35 },
    tint: 0x3a3a3a,
  },
  [UnitKind.Archer]: {
    label: 'Archer',
    icon: '🏹',
    speed: 2.4,
    carry: 0,
    radius: 0.24,
    height: 0.8,
    maxHp: 45,
    attack: 9,
    damageType: DamageType.Pierce,
    armorClass: ArmorClass.Leather,
    range: 5,
    attackRate: 1.4,
    aggroRange: 7,
    cost: { wood: 45 },
    tint: 0x5a3a1a,
  },
  [UnitKind.Knight]: {
    label: 'Knight',
    icon: '🐎',
    speed: 3.4, // fast — runs down archers
    carry: 0,
    radius: 0.3,
    height: 1.0,
    maxHp: 130,
    attack: 17,
    damageType: DamageType.Slash, // shreds unarmored/leather, glances mail
    armorClass: ArmorClass.Mail, // shrugs off arrows, but spears punch through
    range: 1.0,
    attackRate: 1.1,
    aggroRange: 7,
    cost: { wood: 90 },
    tint: 0x9a8050,
    requires: BuildingKind.Stable, // moved out of the Barracks into the Stable
  },
  [UnitKind.HorseArcher]: {
    label: 'Horse Archer',
    icon: '🏇',
    speed: 4.0, // fastest on the field — kites everything
    carry: 0,
    radius: 0.28,
    height: 0.95,
    maxHp: 60, // fragile for cavalry
    attack: 8,
    damageType: DamageType.Pierce,
    armorClass: ArmorClass.Leather,
    range: 4.5,
    attackRate: 1.3,
    aggroRange: 8,
    cost: { wood: 40, gold: 20 },
    tint: 0x7a5a2a,
    requires: BuildingKind.Stable,
  },
  [UnitKind.Mamluk]: {
    label: 'Mamluk',
    icon: '🗡️',
    speed: 3.6,
    carry: 0,
    radius: 0.31,
    height: 1.05,
    maxHp: 150, // elite — tougher than a knight
    attack: 19,
    damageType: DamageType.Slash, // strong vs leather/unarmored
    armorClass: ArmorClass.Mail,
    bonusVsArmor: { [ArmorClass.Leather]: 1.4 }, // hunts archers/spearmen
    range: 1.0,
    attackRate: 1.0,
    aggroRange: 7,
    cost: { food: 60, gold: 50 },
    tint: 0xc9a24a,
    requires: BuildingKind.Stable,
  },
  [UnitKind.Crossbowman]: {
    label: 'Crossbowman',
    icon: '🎯',
    speed: 2.0, // slow, heavy
    carry: 0,
    radius: 0.25,
    height: 0.82,
    maxHp: 55,
    attack: 14,
    damageType: DamageType.Pierce,
    armorClass: ArmorClass.Leather,
    bonusVsArmor: { [ArmorClass.Mail]: 2.2 }, // bolts punch mail — counters knights at range
    range: 5.5,
    attackRate: 2.0, // slow reload
    aggroRange: 7,
    cost: { wood: 40, gold: 20 },
    tint: 0x4a3a2a,
  },
  [UnitKind.Ram]: {
    label: 'Battering Ram',
    icon: '🪵',
    speed: 1.2, // crawls
    carry: 0,
    radius: 0.5,
    height: 1.1,
    maxHp: 400, // very high hp — soaks fire while it works the gate
    attack: 40,
    damageType: DamageType.Siege,
    armorClass: ArmorClass.Mail, // timber roof shrugs arrows, melee chips it
    range: 1.5,
    attackRate: 2.4,
    aggroRange: 0, // never auto-aggro — driven onto targets
    cost: { wood: 120 },
    tint: 0x6b4a2b,
    requires: BuildingKind.SiegeWorkshop,
    prefersBuildings: true,
  },
  [UnitKind.Mangonel]: {
    label: 'Mangonel',
    icon: '💥',
    speed: 1.0,
    carry: 0,
    radius: 0.45,
    height: 1.0,
    maxHp: 90, // fragile — must be screened
    attack: 30,
    damageType: DamageType.Siege,
    armorClass: ArmorClass.Unarmored,
    range: 8, // long siege reach
    attackRate: 3.0,
    aggroRange: 9,
    cost: { wood: 100, gold: 60 },
    tint: 0x5a4632,
    requires: BuildingKind.SiegeWorkshop,
    prefersBuildings: true,
  },
};

File: saladin-bevy/crates/sim/src/units.rs
use crate::economy::ResourceCost;
use crate::enums::{ArmorClass, BuildingKind, DamageType, UnitKind};
use crate::math::{Fx, ONE};

/// Stats + presentation for one trainable unit. The generic combat/gather/move
/// systems dispatch on `UnitKind`, so a new roster entry never touches systems.
#[derive(Clone, Copy, Debug)]
pub struct UnitDef {
    pub label: &'static str,
    pub icon: &'static str,
    pub speed: Fx,
    pub carry: i32,
    pub radius: Fx,
    pub height: Fx,
    pub max_hp: i32,
    pub attack: i32,
    pub damage_type: DamageType,
    pub armor_class: ArmorClass,
    /// Specialist multiplier vs each armor class (1.0 == none).
    pub bonus_vs_armor: [Fx; 4],
    pub range: Fx,
    pub attack_rate: Fx,
    pub aggro_range: Fx,
    pub cost: ResourceCost,
    pub tint: Option<u32>,
    pub requires: Option<BuildingKind>,
    pub prefers_buildings: bool,
    pub morale_aura: Fx,
    pub garrisonable: bool,
    pub ranged: bool,
}

const NO_BONUS: [Fx; 4] = [ONE; 4];

/// `bonus_vs_armor` array with one override at `ArmorClass` index.
const fn bonus(ac: ArmorClass, mult: Fx) -> [Fx; 4] {
    let mut b = [ONE; 4];
    b[ac as usize] = mult;
    b
}

const UNIT_DEFS: [UnitDef; 10] = [
    // Peasant
    UnitDef {
        label: "Peasant",
        icon: "🧑‍🌾",
        speed: crate::fx!("2.5"),
        carry: 8,
        radius: crate::fx!("0.22"),
        height: crate::fx!("0.7"),
        max_hp: 30,
        attack: 0,
        damage_type: DamageType::Blunt,
        armor_class: ArmorClass::Unarmored,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("0.8"),
        attack_rate: crate::fx!("1.2"),
        aggro_range: crate::fx!("0"),
        cost: ResourceCost::new(20, 0, 0, 0),
        tint: None,
        requires: None,
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: false,
    },
    // Spearman
    UnitDef {
        label: "Spearman",
        icon: "🛡️",
        speed: crate::fx!("2.2"),
        carry: 0,
        radius: crate::fx!("0.26"),
        height: crate::fx!("0.85"),
        max_hp: 70,
        attack: 12,
        damage_type: DamageType::Pierce,
        armor_class: ArmorClass::Leather,
        bonus_vs_armor: bonus(ArmorClass::Mail, crate::fx!("2.6")),
        range: crate::fx!("1.2"),
        attack_rate: crate::fx!("1.0"),
        aggro_range: crate::fx!("6"),
        cost: ResourceCost::new(35, 0, 0, 0),
        tint: Some(0x3a3a3a),
        requires: None,
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: true,
        ranged: false,
    },
    // Archer
    UnitDef {
        label: "Archer",
        icon: "🏹",
        speed: crate::fx!("2.4"),
        carry: 0,
        radius: crate::fx!("0.24"),
        height: crate::fx!("0.8"),
        max_hp: 45,
        attack: 9,
        damage_type: DamageType::Pierce,
        armor_class: ArmorClass::Leather,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("5"),
        attack_rate: crate::fx!("1.4"),
        aggro_range: crate::fx!("7"),
        cost: ResourceCost::new(45, 0, 0, 0),
        tint: Some(0x5a3a1a),
        requires: None,
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: true,
        ranged: true,
    },
    // Knight
    UnitDef {
        label: "Knight",
        icon: "🐎",
        speed: crate::fx!("3.4"),
        carry: 0,
        radius: crate::fx!("0.3"),
        height: crate::fx!("1.0"),
        max_hp: 130,
        attack: 17,
        damage_type: DamageType::Slash,
        armor_class: ArmorClass::Mail,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("1.0"),
        attack_rate: crate::fx!("1.1"),
        aggro_range: crate::fx!("7"),
        cost: ResourceCost::new(90, 0, 0, 0),
        tint: Some(0x9a8050),
        requires: Some(BuildingKind::Stable),
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: false,
    },
    // HorseArcher
    UnitDef {
        label: "Horse Archer",
        icon: "🏇",
        speed: crate::fx!("4.0"),
        carry: 0,
        radius: crate::fx!("0.28"),
        height: crate::fx!("0.95"),
        max_hp: 60,
        attack: 8,
        damage_type: DamageType::Pierce,
        armor_class: ArmorClass::Leather,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("4.5"),
        attack_rate: crate::fx!("1.3"),
        aggro_range: crate::fx!("8"),
        cost: ResourceCost::new(40, 0, 0, 20),
        tint: Some(0x7a5a2a),
        requires: Some(BuildingKind::Stable),
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: true,
    },
    // Mamluk
    UnitDef {
        label: "Mamluk",
        icon: "🗡️",
        speed: crate::fx!("3.6"),
        carry: 0,
        radius: crate::fx!("0.31"),
        height: crate::fx!("1.05"),
        max_hp: 150,
        attack: 19,
        damage_type: DamageType::Slash,
        armor_class: ArmorClass::Mail,
        bonus_vs_armor: bonus(ArmorClass::Leather, crate::fx!("1.4")),
        range: crate::fx!("1.0"),
        attack_rate: crate::fx!("1.0"),
        aggro_range: crate::fx!("7"),
        cost: ResourceCost::new(0, 0, 60, 50),
        tint: Some(0xc9a24a),
        requires: Some(BuildingKind::Stable),
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: false,
    },
    // Crossbowman
    UnitDef {
        label: "Crossbowman",
        icon: "🎯",
        speed: crate::fx!("2.0"),
        carry: 0,
        radius: crate::fx!("0.25"),
        height: crate::fx!("0.82"),
        max_hp: 55,
        attack: 14,
        damage_type: DamageType::Pierce,
        armor_class: ArmorClass::Leather,
        bonus_vs_armor: bonus(ArmorClass::Mail, crate::fx!("2.2")),
        range: crate::fx!("5.5"),
        attack_rate: crate::fx!("2.0"),
        aggro_range: crate::fx!("7"),
        cost: ResourceCost::new(40, 0, 0, 20),
        tint: Some(0x4a3a2a),
        requires: None,
        prefers_buildings: false,
        morale_aura: crate::fx!("0"),
        garrisonable: true,
        ranged: true,
    },
    // Ram
    UnitDef {
        label: "Battering Ram",
        icon: "🪵",
        speed: crate::fx!("1.2"),
        carry: 0,
        radius: crate::fx!("0.5"),
        height: crate::fx!("1.1"),
        max_hp: 400,
        attack: 40,
        damage_type: DamageType::Siege,
        armor_class: ArmorClass::Mail,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("1.5"),
        attack_rate: crate::fx!("2.4"),
        aggro_range: crate::fx!("0"),
        cost: ResourceCost::new(120, 0, 0, 0),
        tint: Some(0x6b4a2b),
        requires: Some(BuildingKind::SiegeWorkshop),
        prefers_buildings: true,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: false,
    },
    // Mangonel
    UnitDef {
        label: "Mangonel",
        icon: "💥",
        speed: crate::fx!("1.0"),
        carry: 0,
        radius: crate::fx!("0.45"),
        height: crate::fx!("1.0"),
        max_hp: 90,
        attack: 30,
        damage_type: DamageType::Siege,
        armor_class: ArmorClass::Unarmored,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("8"),
        attack_rate: crate::fx!("3.0"),
        aggro_range: crate::fx!("9"),
        cost: ResourceCost::new(100, 0, 0, 60),
        tint: Some(0x5a4632),
        requires: Some(BuildingKind::SiegeWorkshop),
        prefers_buildings: true,
        morale_aura: crate::fx!("0"),
        garrisonable: false,
        ranged: false,
    },
    // Imam
    UnitDef {
        label: "Imam",
        icon: "🕌",
        speed: crate::fx!("2.6"),
        carry: 0,
        radius: crate::fx!("0.24"),
        height: crate::fx!("0.85"),
        max_hp: 50,
        attack: 0,
        damage_type: DamageType::Blunt,
        armor_class: ArmorClass::Unarmored,
        bonus_vs_armor: NO_BONUS,
        range: crate::fx!("0"),
        attack_rate: crate::fx!("0"),
        aggro_range: crate::fx!("0"),
        cost: ResourceCost::new(0, 0, 40, 0),
        tint: Some(0xe8e2d0),
        requires: None,
        prefers_buildings: false,
        morale_aura: crate::fx!("7"),
        garrisonable: true,
        ranged: false,
    },
];

pub fn unit_def(kind: UnitKind) -> &'static UnitDef {
    &UNIT_DEFS[kind as usize]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_kind_indexes() {
        for &k in UnitKind::ALL {
            assert_eq!(unit_def(k).label.is_empty(), false);
        }
    }

    #[test]
    fn spearman_braces_against_mail() {
        let d = unit_def(UnitKind::Spearman);
        assert_eq!(d.bonus_vs_armor[ArmorClass::Mail as usize], crate::fx!("2.6"));
        assert_eq!(d.bonus_vs_armor[ArmorClass::Leather as usize], ONE);
    }

    #[test]
    fn cavalry_requires_stable() {
        assert_eq!(unit_def(UnitKind::Knight).requires, Some(BuildingKind::Stable));
        assert_eq!(unit_def(UnitKind::Peasant).requires, None);
    }
}
